src/lib/crypto/sol-message-digest-impl-openssl.c

```diff
@@ -0,0 +1,780 @@
+_sol_message_digest_evp_update(struct sol_message_digest *handle, struct sol_blob *input)
+{
+    return !!EVP_DigestUpdate(&handle->ctx.evp, input->mem, input->size);
+}
+
+static bool
+_sol_message_digest_evp_final(struct sol_message_digest *handle, struct sol_blob *digest)
+{
+    unsigned int len = digest->size;
+
+    if (EVP_DigestFinal_ex(&handle->ctx.evp, digest->mem, &len)) {
+        if (len != digest->size) {
+            SOL_WRN("Wanted %zd digest bytes, got %u", digest->size, len);
+            digest->size = len;
+        }
+        return true;
+    }
+    return false;
+}
+
+static const struct sol_message_digest_openssl_ops _sol_message_digest_openssl_ops_evp = {
+    .init = _sol_message_digest_evp_init,
+    .cleanup = _sol_message_digest_evp_cleanup,
+    .update = _sol_message_digest_evp_update,
+    .final = _sol_message_digest_evp_final
+};
+
+static bool
+_sol_message_digest_hmac_init(struct sol_message_digest *handle, const EVP_MD *md, const struct sol_str_slice key)
+{
+    return !!HMAC_Init(&handle->ctx.hmac, key.data, key.len, md);
+}
+
+static void
+_sol_message_digest_hmac_cleanup(struct sol_message_digest *handle)
+{
+    HMAC_CTX_cleanup(&handle->ctx.hmac);
+}
+
+static bool
+_sol_message_digest_hmac_update(struct sol_message_digest *handle, struct sol_blob *input)
+{
+    return !!HMAC_Update(&handle->ctx.hmac, input->mem, input->size);
+}
+
+static bool
+_sol_message_digest_hmac_final(struct sol_message_digest *handle, struct sol_blob *digest)
+{
+    unsigned int len = digest->size;
+
+    if (HMAC_Final(&handle->ctx.hmac, digest->mem, &len)) {
+        if (len != digest->size) {
+            SOL_WRN("Wanted %zd digest bytes, got %u", digest->size, len);
+            digest->size = len;
+        }
+        return true;
+    }
+    return false;
+}
+
+static const struct sol_message_digest_openssl_ops _sol_message_digest_openssl_ops_hmac = {
+    .init = _sol_message_digest_hmac_init,
+    .cleanup = _sol_message_digest_hmac_cleanup,
+    .update = _sol_message_digest_hmac_update,
+    .final = _sol_message_digest_hmac_final
+};
+
+SOL_API struct sol_message_digest *
+sol_message_digest_new(const struct sol_message_digest_config *config)
+{
+    const struct sol_message_digest_openssl_ops *ops = NULL;
+    const EVP_MD *md;
+    struct sol_message_digest *handle;
+    int errno_bkp;
+    bool r;
+
+    errno = EINVAL;
+    SOL_NULL_CHECK(config, NULL);
+    SOL_NULL_CHECK(config->on_digest_ready, NULL);
+    SOL_NULL_CHECK(config->algorithm, NULL);
+
+    if (config->api_version != SOL_MESSAGE_DIGEST_CONFIG_API_VERSION) {
+        SOL_WRN("sol_message_digest_config->api_version=%hu, "
+            "expected version is %hu.",
+            config->api_version, SOL_MESSAGE_DIGEST_CONFIG_API_VERSION);
+        return NULL;
+    }
+
+    if (!did_openssl_load_digests) {
+        OpenSSL_add_all_digests();
+        did_openssl_load_digests = true;
+    }
+
+    md = EVP_get_digestbyname(config->algorithm);
```

Comment:
Wouldn't it be a good idea to standardize the digest names in order to keep the uses portable? Or are the strings exported by `kcapi` the same supported by `libcrypto`?

Maybe a `sol_str_table` with a fallback to NULL, where it would then be sent to the implementation to decide?